include/deal.II/physics/notation.h

```diff
@@ -0,0 +1,1513 @@
+            Assert(indices.second == 0, ExcInternalError());
+            const unsigned int &i = indices.first;
+
+            for (unsigned int c=0; c<n_cols; ++c)
+              {
+                Assert(c < 1, ExcInternalError());
+                v[i] = mtrx(r,c);
+              }
+          }
+      }
+
+
+      template<int dim, typename Number>
+      void
+      roll_to_tensor (Tensor<2,dim,Number>     &t,
+                      const FullMatrix<Number> &mtrx)
+      {
+        Assert(mtrx.m() == dim, ExcDimensionMismatch(mtrx.m(), dim));
+        Assert(mtrx.n() == dim, ExcDimensionMismatch(mtrx.n(), dim));
+        Assert(mtrx.n_elements() == t.n_independent_components,
+               ExcMessage("Input matrix has incorrect number of entries."));
+
+        const unsigned int n_rows = mtrx.m();
+        const unsigned int n_cols = mtrx.n();
+        for (unsigned int r=0; r<n_rows; ++r)
+          {
+            const std::pair<unsigned int,unsigned int> indices_i = internal::indices_from_component<dim>(r,false);
+            Assert(indices_i.first < dim, ExcInternalError());
+            Assert(indices_i.second < dim, ExcInternalError());
+            const unsigned int &i = indices_i.first;
+
+            for (unsigned int c=0; c<n_cols; ++c)
+              {
+                const std::pair<unsigned int,unsigned int> indices_j = internal::indices_from_component<dim>(c,false);
+                Assert(indices_j.first < dim, ExcInternalError());
+                Assert(indices_j.second < dim, ExcInternalError());
+                const unsigned int &j = indices_j.second;
+
+                t[i][j] = mtrx(r,c);
+              }
+          }
+
+      }
+
+
+      template<int dim, typename Number>
+      void
+      roll_to_tensor (SymmetricTensor<2,dim,Number> &st,
+                      const FullMatrix<Number>      &mtrx)
+      {
+        // Its impossible to fit the (dim^2 + dim)/2 entries into a square matrix
+        // We therefore assume that its been converted to a standard tensor format
+        // using unroll_to_matrix (SymmetricTensor<2,dim,Number>) at some point...
+        Assert(mtrx.m() == dim, ExcDimensionMismatch(mtrx.m(), dim));
+        Assert(mtrx.n() == dim, ExcDimensionMismatch(mtrx.n(), dim));
+        Assert((mtrx.n_elements() == Tensor<2,dim,Number>::n_independent_components),
+               ExcMessage("Input matrix has incorrect number of entries."));
+
+        Tensor<2,dim,Number> tmp;
+        roll_to_tensor(tmp,mtrx);
+        st = symmetrize(tmp);
+        Assert((Tensor<2,dim,Number>(st) - tmp).norm() < 1e-12,
+               ExcMessage("The entries stored inside the matrix were not symmetric"));
+      }
+
+
+      template<int dim, typename Number>
+      void
+      roll_to_tensor (Tensor<3,dim,Number>     &t,
+                      const FullMatrix<Number> &mtrx)
+      {
+        Assert((mtrx.m() == Tensor<1,dim,Number>::n_independent_components) ||
```

Comment:
what you can do to help you here is declare
```
DeclException4 (NotationExcTripleSize, int, int, int, int,
                  << "Number of rows " << arg1
                  << " shall be one of " << arg2 << "," << arg3 << "," << arg4);
```
and use it
```
NotationExcTripleSize(mtrx.m(), Tensor<1,dim,Number>::n_independent_components, Tensor<2,dim,Number>::n_independent_components, SymmetricTensor<2,dim,Number>::n_independent_components)
```
(of course name and message are arbitrary)

Reply:
same for other complicated asserts.

Reply:
Done. Thanks for the tip.